### PR TITLE
feat(goals): surface money that left a goal's accounts unexplained

### PR DESCRIPTION
**Pull request by buzzromain**
> [!IMPORTANT]
> **Stacked on #3160, #3165, #3167 and #3176 — do not merge before them.** The commit list carries all four. One commit belongs to this PR, `feat(goals): surface money that left a goal's accounts unexplained`, and it rebases down to that single commit once they land. #3176 is a hard dependency: this is the surface that calls its `consume!`.

## Why

Goals never read transactions. `current_balance` is a stock summed from account balances, so an outflow reaches a goal only as a smaller number, with nothing saying which goal it belonged to.

#3176 gives the user a way to say *"I spent this"*. It does not give them a reason to think of it — and that is the whole difficulty. Somebody who has just come home from the holiday their goal paid for is not going to open the app and file it. The app already knows the money left; it was simply waiting to be asked.

## What changed

`Goal::WithdrawalDetector` finds outflows on a goal's accounts that nothing has claimed, and the goal page offers them: *if any of this was spent on Trip, say so*. One click records it, with the transaction as evidence.

This is the pull half of what `GoalPledge` does for money coming in. A pledge asks first and matches later; here there is nothing to promise, so the outflow is surfaced after the fact and attributed — or ignored, which is also an answer.

## What reviewers should look at

**The sign.** In Sure an inflow carries a *negative* amount, so the detector selects the positive side. Reading it the other way round would offer to attribute the user's own deposits as spending, and the mistake would look perfectly reasonable in a diff. There is a test whose only job is to pin that.

**The stamp.** `consume!` now takes a transaction and writes `extra["goal"]["consumed_goal_id"]` — the same namespace the pledges already use, so both halves of a goal's money are marked alike. That is what makes attribution idempotent: replaying it cannot credit a goal twice for one spend. The stamp happens inside the consumption's own database transaction, so a refusal rolls the whole thing back rather than leaving a goal credited for a spend it did not record.

**The reserve exclusion.** A reserve is drawn down and refilled, not spent. Asking someone to attribute a withdrawal from one invites them to erase the shortfall it exists to report — the same signal #3175 raises an insight about.

## Bounds, and why each one

- **`MIN_AMOUNT`** — below it an attribution is more friction than it is worth. A goal-backed savings account sees card-sized outflows nobody wants to file. Same known limitation as `IdleCashGenerator::MIN_BALANCE`: a flat figure in family-currency units, tuned for dollar/euro-scale currencies.
- **`LOOKBACK_DAYS`** — an outflow nobody attributed in three months is one nobody is going to. Asking forever turns the panel into a chore.
- **Three at a time, newest first** — the panel is a prompt, not a reconciliation ledger.

## A known asymmetry

`GoalPledge::Reconciler` only runs on provider imports, never on a hand-entered transaction. This detector reads entries directly and has no such gap, so the two halves of the feature do not behave the same way on a manual account. Worth knowing rather than discovering.

## Testing

`bin/rails test` — 7100 runs, 28540 assertions, 0 failures, 0 errors. RuboCop, erb_lint and Brakeman clean.

Nine detector tests: the happy path, the sign, the threshold, excluded entries, the window, an already-attributed outflow, a reserve, an account the goal does not fund, and the ordering with its cap. Three controller tests: the panel appearing, an attribution taking the transaction's own amount and disappearing afterwards, and the same outflow refused on a second attempt.

## Summary by CodeRabbit

- **New Features**
  - Added one-off goals and maintained reserve goals.
  - Record partial spending and attribute unmatched outflows.
  - Added reserve funding, depletion, shortfall, and intact-status indicators.
  - Goal forms adapt to the selected goal type.
  - Added clearer lifecycle panels and completion confirmations.

- **Improvements**
  - Completed goals retain achieved amounts after spending.
  - Earmarking and projections better reflect released, consumed, and reserved funds.
  - Prevented conflicting whole-account allocations.
  - Added safeguards against duplicate spending and invalid goal changes.
  - Added clearer validation messages and English/French translations.

### CODE REVIEW

File: app/controllers/goals_controller.rb [hunk 1]
```diff
@@ -153,6 +161,22 @@ def unarchive
...
+  # Renders the dialog. The write lives in its own action below.
+  def consume
+  end
+
+  def record_consumption
+    txn = consumption_transaction
+    amount = txn ? txn.entry.amount.to_d : params[:amount]
+
+    @goal.consume!(amount, account: consumption_account(txn), transaction: txn)
+    redirect_to goal_path(@goal), notice: t("goals.consume.success", amount: params[:amount])
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/controllers/goals_controller.rb]: Fixed — and thank you for catching that the form posts only `transaction_id`, which is what made it a blank message rather than a slightly wrong one.

The resolved amount is used now, formatted through `Money.new(amount, @goal.currency)`. Same change also fixes the typed path, where the raw parameter was being interpolated unformatted.

File: app/models/goal/withdrawal_detector.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      Entry
+        .joins("INNER JOIN transactions ON transactions.id = entries.entryable_id AND entries.entryable_type = 'Transaction'")
+        .where(account_id: goal.linked_accounts.map(&:id))
+        .where(excluded: false)
+        .where(date: LOOKBACK_DAYS.days.ago.to_date..Date.current)
+        # In Sure an inflow carries a NEGATIVE amount, so an outflow is the
+        # positive side. Reading this the other way round would offer to
+        # attribute the user's deposits as spending.
+        .where("entries.amount >= ?", MIN_AMOUNT)
+        .where("transactions.extra -> 'goal' ->> 'consumed_goal_id' IS NULL")
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/models/goal/withdrawal_detector.rb:67]: Fixed in f196473b, through `Transaction.pending_providers_sql` so it stays in step with the rest of the app rather than growing its own list of providers.

The failure mode you describe is the worse half: the goal keeps consumption for a transaction that no longer exists, *and* the posted twin arrives unstamped and gets offered again — so the user can attribute the same spend twice and the second one looks perfectly legitimate.

Tests: `ignores an outflow the provider has not posted yet` and `still surfaces an outflow the provider has posted`, the second so the filter cannot quietly swallow everything.

File: app/views/goals/_unattributed_outflows.html.erb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        </div>
+        <div class="flex items-center gap-2 shrink-0">
+          <span class="text-sm text-primary tabular-nums privacy-sensitive">
+            <%= Money.new(entry.amount, entry.currency).format(precision: 0) %>
+          </span>
+          <%= button_to t(".attribute"),
+                        consume_goal_path(goal),
+                        method: :post,
+                        params: { transaction_id: entry.entryable_id },
+                        class: "btn btn--secondary btn--sm" %>
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/views/goals/_unattributed_outflows.html.erb]: Fixed in f196473b — `DS::Button` with `variant: :secondary, size: :sm`. My mistake, and the inconsistency was visible on the same page: the consumption dialog beside it was already using the primitive.

File: app/models/goal/withdrawal_detector.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  def initialize(goal)
+    @goal = goal
+  end
+
+  # Outflows on this goal's accounts that nothing has claimed yet, newest
+  # first. Empty for a goal that cannot consume — a reserve is drawn down and
+  # refilled rather than spent, and its shortfall already says so on its own.
+  def unattributed_outflows(limit: 3)
+    return Entry.none unless goal.one_off?
+    return Entry.none if goal.linked_accounts.empty?
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/models/goal/withdrawal_detector.rb]: Fixed in f196473b — the detector returns nothing for a released goal, and `consume!` refuses one outright (`:not_active`, added on #3176 and carried down here by rebase).

Both halves were needed and you named why: the panel offering the outflow is the invitation, the missing state guard is what let the POST through. Closing only the second would leave the page suggesting an action it then refuses.

Tests: `says nothing once the goal has been released` and `says nothing for an archived goal`.

File: app/models/goal.rb [hunk 5]
```diff
@@ -665,10 +843,52 @@ def pooled_pace
...
+        raise ConsumptionRefused.new(:account_required) if links.size > 1
+        return links.first
+      end
+
+      links.find { |link| link.account_id == account.id } ||
+        raise(ConsumptionRefused.new(:account_not_linked))
+    end
+
+    def thaw_completed_amount!
+      update_columns(completed_amount: nil, completed_at: nil, consumed_amount: 0)
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/models/goal.rb]: Good catch — fixed in f196473b.

You are right that the two paths are not the same event. Restarting a goal is what drops what it spent under its previous life, and a goal archived straight from active never closed a lifecycle to restart from; unarchiving it is picking the same goal back up. `thaw_completed_amount!` now clears `consumed_amount` only when a frozen figure exists.

Tests: `unarchiving a goal that never completed keeps what it spent`, and `reopening a completed goal starts it over` so the original behaviour stays pinned.

File: app/models/goal.rb [hunk 6]
```diff
@@ -226,8 +335,56 @@ def market_value_money
...
+      link.lock!
+      stamp_consumption!(transaction) if transaction
+
+      # A whole-account link reserves no fixed slice, so there is nothing to
+      # shrink — it already takes only what the account has left.
+      if link.allocated_amount.present?
+        link.update!(allocated_amount: [ link.allocated_amount.to_d - amount, 0 ].max)
+      end
+
+      update!(consumed_amount: consumed_amount.to_d + amount)
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/models/goal.rb:460]: Fixed on #3176 and carried down here by rebase. `consumed_amount` lives on the goal, so locking only the link left both callers holding the same stale value — the target check included. The whole check-and-increment now runs inside `with_lock` on the goal, with the link locked underneath it, consistently goal-first so the two orderings cannot deadlock.

File: app/controllers/goals_controller.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/controllers/goals_controller.rb]: Fixed — the resolved amount, formatted with the goal's currency, on both paths. The one-click attribution produced a blank message because the form sends only `transaction_id`; the typed path was interpolating the raw parameter.

File: app/models/goal.rb [hunk 7]
```diff
@@ -226,8 +335,56 @@ def market_value_money
...
+    end
+
+    reload
   end
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/models/goal.rb:469]: Fixed on #3176 and carried down here by rebase — `with_lock` on the goal wraps the target check and the increment, so the value is re-read after waiting rather than before.

Your note that recording again recovers the amount is right, and it is what makes this quieter than it looks: the user sees a goal that did not move as much as they expected and no error, which is harder to notice than a failure would be.

File: app/controllers/goals_controller.rb [hunk 8]
```diff
@@ -153,6 +161,22 @@ def unarchive
...
 
+  # Renders the dialog. The write lives in its own action below.
+  def consume
+  end
+
+  def record_consumption
+    txn = consumption_transaction
+    amount = txn ? txn.entry.amount.to_d : params[:amount]
+
+    @goal.consume!(amount, account: consumption_account(txn), transaction: txn)
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/controllers/goals_controller.rb:173]: Fixed. `consume!` now raises `:not_active` inside the goal lock (#3176, carried down by rebase), with a localized message, and the detector returns no candidates for a released goal (f196473b).

One narrowing worth flagging: the refusal covers `completed`, `archived` and `paused` alike, since `active?` is the condition. Pausing is the user saying "not now", so accepting a spend against a paused goal seemed wrong for the same reason — but it is a judgement, and I would change it if you read pause as merely cosmetic.

Coverage is per-behaviour rather than per-state: `a goal that is no longer active records nothing` on the model, plus the two detector tests.

**Comment by buzzromain** [comment]
One change on this branch that no thread asked for, flagged here because it is security-relevant and easy to miss in the diff.

`consumption_account` resolved the account behind an attributed outflow through `Current.family.accounts`, while the named-account path was moved to `Current.user.accessible_accounts` on #3176. That left the hole open on this branch through the other door: a family member attributing an outflow on a private account they cannot see would have its earmark reduced, and learn the account exists from the figures that moved.

Both paths now resolve through `accessible_accounts`, and an id that does not resolve is refused rather than falling through to `nil`.

File: app/components/goals/lifecycle_panel_component.html.erb [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        </span>
+        <% if show_projection_legend? %>
+          <span class="inline-flex items-center gap-1.5">
+            <svg width="18" height="6"><line x1="0" y1="3" x2="18" y2="3" stroke="<%= projection_color %>" stroke-width="2" stroke-dasharray="3 3" /></svg>
+            <%= t("goals.show.projection.legend_projection") %>
+          </span>
+          <% if show_required_legend? %>
+            <span class="inline-flex items-center gap-1.5">
+              <svg width="18" height="6" class="text-secondary"><line x1="0" y1="3" x2="18" y2="3" stroke="currentColor" stroke-width="2" stroke-dasharray="2 4" opacity="0.5" /></svg>
+              <%= t("goals.show.projection.legend_required") %>
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/components/goals/lifecycle_panel_component.html.erb]: Leaving this one, and flagging the judgement rather than quietly ignoring it.

That legend markup is not new — it moved across verbatim from `goals/show.html.erb` in the extraction this PR does, and the extraction was itself a review request to get the decision logic out of the template. Rewriting the markup in the same commit would have made "did anything change visually?" unanswerable from the diff, which is the one question that extraction has to be able to answer cheaply.

There is also no DS primitive for a legend swatch — `app/components/DS` has no candidate — so the fix is "add a primitive", which is a change worth making on its own terms rather than as a rider here.

Same for `text-[11px]`: it is the existing value, and `text-xs` is 12px, so swapping it is a visual change rather than a token cleanup.

Glad to do both as a follow-up if you want them, and happy to do them here instead if you would rather not split it.

File: test/components/goals/lifecycle_panel_component_test.rb [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+      ) { |g| g.goal_accounts.build(account: pot(balance: 1_000), allocated_amount: 1_000) }
+    end
+
+    def component_for(goal)
+      Goals::LifecyclePanelComponent.new(goal: goal)
+    end
+
+    def panel_for(goal)
+      component_for(goal).panel
+    end
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply test/components/goals/lifecycle_panel_component_test.rb:73]: I do not think this one holds. `Layout/IndentationConsistency` in this repo is on the `indented_internal_methods` setting, so four spaces after `private` is what RuboCop wants here — `bin/rubocop` passes on this file and would flag two.

It is also what the rest of the suite does: of the test files under `test/models` with a `private` section, 14 indent the following methods by four spaces and none by two.

Happy to change it if the project would rather move the other way, but it would be a sweep rather than this file.

File: app/models/goal.rb [hunk 11]
```diff
@@ -761,10 +1055,91 @@ def linked_accounts_must_belong_to_family
...
+
+    # The consumed total is checked while consuming, which left the ordinary
+    # edit form free to lower the target underneath it — a goal reporting more
+    # spent than it ever set out to save.
+    def target_must_cover_what_was_consumed
+      return unless target_amount.present? && consumed_amount.to_d.positive?
+      return if target_amount.to_d >= consumed_amount.to_d
+
+      errors.add(:target_amount, :below_consumed)
+    end
```
[the remark this answers is not in the document]
Comment (buzzromain) [reply app/models/goal.rb:1131]: Same answer as on #3176, repeated here so this thread stands on its own.

The nine locales are already several goal keys behind on `main` — this is drift across the whole stack, not something this PR introduces. `config.i18n.fallbacks = true` is set in both `application.rb` and `production.rb`, and there is no `i18n-tasks` check in CI, so English fallback is the designed behaviour here rather than a break.

I would rather not machine-translate financial error messages into nine languages I cannot review. "This goal has already recorded money as spent" being subtly wrong in Polish is worse for the reader than the English they would otherwise see, and harder to spot afterwards because the key now looks filled in.

Glad to fill them in if the project wants — I would just do it as one pass over the whole goal namespace, and would want to know how translations normally reach this repo.

**Comment by sure-design** [comment]
## DS Drift Patrol — 1 finding

### Rule 5 — i18n bypass
- `app/views/goals/_unattributed_outflows.html.erb` — the `·` separator between the entry date and account name is a bare literal, not wrapped in `t()`:
  ```erb
  <%= l(entry.date, format: :long) %> · <%= entry.account.name %>
  ```
  This is the same hardcoded-separator pattern flagged repeatedly elsewhere in this repo (#3034, #3054, and three merged commits this week — see the `ds-drift` tracking issue). Suggested fix: a shared locale key (e.g. `shared.separator`) or a small helper, since this keeps recurring across unrelated PRs.

Everything else in this PR's in-scope surface (`app/views/goals/consume.html.erb`, `app/views/goals/show.html.erb`) is clean — `DS::Dialog`, `DS::Select`, `DS::Button` used correctly, functional color tokens throughout.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by buzzromain** [comment]
Fixed — and you were right that the mechanism already exists: `shared.dot_separator` is defined in every shipped locale and already used three times in the goals views (`goals/index.html.erb` twice, `goals/_status_callout.html.erb`). This was drift, not a gap.

Worth noting what the existing uses do that the literal did not: they wrap the separator in `aria-hidden="true"`. The `·` is decorative, and a screen reader was reading it out between the date and the account name. The fix follows that shape rather than only the `t()` call.

On the recurrence you flag — I checked the rest of the goals surface and there are six more hardcoded `·`, all in files this PR does not touch and all already on `main`:

- `app/views/goals/_pending_pledge_banner.html.erb`
- `app/views/goals/index.html.erb` (a `parts.join(" · ")`)
- `app/views/goals/show.html.erb` (a `safe_join` delimiter)
- `app/views/goals/_form.html.erb`
- `app/components/goals/card_component.html.erb`
- `app/components/goals/funding_accounts_breakdown_component.html.erb`

I deliberately left them. `show.html.erb` is in this PR's diff, but the separator line is pre-existing and untouched by it — folding an unrelated cleanup into a feature PR makes the diff harder to review for no gain, and the other five are in files this PR does not open at all.

Two of those are join delimiters rather than inline markup, so they need slightly different handling than a `<span aria-hidden>`, which is another reason to do them together rather than one at a time. Happy to send that as its own PR across the goals surface if a maintainer wants it — it would close the recurrence properly rather than one file per unrelated PR.

**Comment by jjmata** [comment]
> Two of those are join delimiters rather than inline markup, so they need slightly different handling than a `<span aria-hidden>`, which is another reason to do them together rather than one at a time. Happy to send that as its own PR across the goals surface if a maintainer wants it — it would close the recurrence properly rather than one file per unrelated PR.

Happy to take that as a separate PR, yes!

**Review by jjmata — APPROVED** [review]
